segmentstore/server/src/test/java/io/pravega/segmentstore/server/DataRecoveryTestUtils.java

```diff
@@ -0,0 +1,188 @@
+            int containerId = segToConMapper.getContainerId(curr.getName());
+            List<SegmentProperties> segmentsList = segmentToContainers.get(containerId);
+            if (segmentsList == null) {
+                segmentsList = new ArrayList<>();
+                segmentsList.add(curr);
+                segmentToContainers.put(containerId, segmentsList);
+            } else {
+                segmentToContainers.get(containerId).add(curr);
+            }
+        }
+        return segmentToContainers;
+    }
+
+    public static ScheduledExecutorService createExecutorService(int threadPoolSize) {
+        ScheduledThreadPoolExecutor es = new ScheduledThreadPoolExecutor(threadPoolSize);
+        es.setContinueExistingPeriodicTasksAfterShutdownPolicy(false);
+        es.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+        es.setRemoveOnCancelPolicy(true);
+        return es;
+    }
+
+     /**
+     * Creates all segments given in the list with the given DebugStreamSegmentContainer.
+     */
+     public static class Worker implements Runnable {
+        private final int containerId;
+        private final DebugStreamSegmentContainer container;
+        private final List<SegmentProperties> segments;
+        public Worker(DebugStreamSegmentContainer container, List<SegmentProperties> segments) {
+            this.container = container;
+            this.containerId = container.getId();
+            this.segments = segments;
+        }
+
+        @Override
+        public void run() {
+            if (segments == null) {
+                return;
+            }
+            log.info("Recovery started for container = {}", containerId);
+            ContainerTableExtension ext = container.getExtension(ContainerTableExtension.class);
+            AsyncIterator<IteratorItem<TableKey>> it = ext.keyIterator(getMetadataSegmentName(containerId),
+                    IteratorArgs.builder().fetchTimeout(TIMEOUT).build()).join();
+
+            // Add all segments present in the container metadata in a set.
+            Set<TableKey> segmentsInMD = new HashSet<>();
+            it.forEachRemaining(k -> segmentsInMD.addAll(k.getEntries()), EXECUTOR_SERVICE).join();
+
+            for (SegmentProperties segment : segments) {
+                long len = segment.getLength();
+                boolean isSealed = segment.isSealed();
+                String segmentName = segment.getName();
+
+                /*
+                    1. segment exists in both metadata and storage, re-create it
+                    2. segment only in metadata, delete
+                    3. segment only in storage, re-create it
+                 */
+                segmentsInMD.remove(TableKey.unversioned(getTableKey(segmentName)));
+                container.getStreamSegmentInfo(segment.getName(), TIMEOUT)
+                        .thenAccept(e -> {
+                            container.createStreamSegment(segmentName, len, isSealed)
```

Comment:
Why do you attempt to create the segment if it already exists? `getStreamSegmentInfo` will return a value (not an exception) if it exists.

Reply:
Now, deleting the segment to remove it from Tier1, and then recreating it using the length and sealed status from LTS.